Openvino 2024.2.0 (#35)
ilya-lavrenov/vllm: docs/source/getting_started/openvino-installation.rst

.. _installation_openvino:

> Installation with OpenVINO
> ========================

Reviewer: ```suggestion
Installation with OpenVINO
==========================
```

> vLLM powered by OpenVINO supports all LLM models from [vLLM supported models list](../dev/models/supported_models.rst) and can perform optimal model serving on all x86-64 CPUs with, at least, AVX2 support. OpenVINO vLLM backend supports the following advanced vLLM features:

Reviewer: ```suggestion
vLLM powered by OpenVINO supports all LLM models from 
:doc:`vLLM supported models list <../models/supported_models>` and can perform optimal model serving on all x86-64 CPUs with at least AVX2 support. OpenVINO vLLM backend supports the following advanced vLLM features:
```

- Prefix caching (``--enable-prefix-caching``)
- Chunked prefill (``--enable-chunked-prefill``)

> Table of contents:
> 
> #. :ref:`Requirements <openvino_backend_requirements>`
> #. :ref:`Quick start using Dockerfile <openvino_backend_quick_start_dockerfile>`
> #. :ref:`Build from source <binstall_openvino_backend_from_source>`
> #. :ref:`Performance tips <openvino_backend_performance_tips>`
> #. :ref:`Limitations <openvino_backend_limitations>`

Reviewer: ```suggestion
**Table of contents**:

- :ref:`Requirements <openvino_backend_requirements>`
- :ref:`Quick start using Dockerfile <openvino_backend_quick_start_dockerfile>`
- :ref:`Build from source <install_openvino_backend_from_source>`
- :ref:`Performance tips <openvino_backend_performance_tips>`
- :ref:`Limitations <openvino_backend_limitations>`
```

.. _openvino_backend_requirements:

Requirements
------------

* OS: Linux
* Instruction set architecture (ISA) requirement: at least AVX2.

.. _openvino_backend_quick_start_dockerfile:

Quick start using Dockerfile
----------------------------

.. code-block:: console

    $ docker build -f Dockerfile.openvino -t vllm-openvino-env .
    $ docker run -it --rm vllm-openvino-env

.. _install_openvino_backend_from_source:

> Install from source
> -----------------

Reviewer: ```suggestion
Install from source
-------------------
```

- First, install Python. For example, on Ubuntu 22.04, you can run:

> .. code-block:: console
> 
>     $ sudo apt-get update  -y
>     $ sudo apt-get install python3

Reviewer: ```suggestion
  .. code-block:: console

      $ sudo apt-get update  -y
      $ sudo apt-get install python3
```

> - Second, install prerequisites vLLM OpenVINO backend installation:

Reviewer: ```suggestion
- Then, install the prerequisites for vLLM OpenVINO backend installation:
```

> .. code-block:: console
> 
>     $ pip install --upgrade pip
>     $ pip install -r requirements-build.txt --extra-index-url https://download.pytorch.org/whl/cpu

Reviewer: ```suggestion
  .. code-block:: console

      $ pip install --upgrade pip
      $ pip install -r requirements-build.txt --extra-index-url https://download.pytorch.org/whl/cpu
```

> - Finally, install vLLM with OpenVINO backend: 

Reviewer: ```suggestion
- Finally, install vLLM OpenVINO backend: 
```

> .. code-block:: console
> 
>     $ PIP_PRE=1 PIP_EXTRA_INDEX_URL="https://download.pytorch.org/whl/cpu https://storage.openvinotoolkit.org/simple/wheels/nightly/" VLLM_TARGET_DEVICE=openvino python install -v .

Reviewer: ```suggestion
  .. code-block:: console

      $ PIP_PRE=1 PIP_EXTRA_INDEX_URL="https://download.pytorch.org/whl/cpu https://storage.openvinotoolkit.org/simple/wheels/nightly/" VLLM_TARGET_DEVICE=openvino python install -v .
```

.. _openvino_backend_performance_tips:

Performance tips
-----------------

> vLLM OpenVINO backend uses the following environment variables to control behavior:

Reviewer: ```suggestion
To control behavior in vLLM OpenVINO backend, use the following environment variables:
```

> - ``VLLM_OPENVINO_KVCACHE_SPACE`` to specify the KV Cache size (e.g, ``VLLM_OPENVINO_KVCACHE_SPACE=40`` means 40 GB space for KV cache), larger setting will allow vLLM running more requests in parallel. This parameter should be set based on the hardware configuration and memory management pattern of users.

Reviewer: ```suggestion
- ``VLLM_OPENVINO_KVCACHE_SPACE`` specifies the KV Cache size (e.g, ``VLLM_OPENVINO_KVCACHE_SPACE=40`` means 40 GB space for KV cache). Higher setting will enable vLLM to run more requests in parallel. This parameter should be set based on the hardware configuration and user-defined memory management pattern.
```

> - ``VLLM_OPENVINO_CPU_KV_CACHE_PRECISION=u8`` to control KV cache precision. By default, FP16 / BF16 is used depending on platform.

Reviewer: ```suggestion
- ``VLLM_OPENVINO_CPU_KV_CACHE_PRECISION=u8`` controls KV cache precision. By default, ``FP16`` / ``BF16`` is used, depending on platform.
```

> - ``VLLM_OPENVINO_ENABLE_QUANTIZED_WEIGHTS=ON`` to enable U8 weights compression during model loading stage. By default, compression is turned off.

Reviewer: ```suggestion
- ``VLLM_OPENVINO_ENABLE_QUANTIZED_WEIGHTS=ON`` enables U8 weights compression during a model loading stage. By default, the compression is turned off.
```

To enable better TPOT / TTFT latency, you can use vLLM's chunked prefill feature (``--enable-chunked-prefill``). Based on the experiments, the recommended batch size is ``256`` (``--max-num-batched-tokens``)

> OpenVINO best known configuration is:

Reviewer: ```suggestion
Best known configuration in OpenVINO is:
```
